controllers/tc000044_plan_only_mode_test.go

```diff
@@ -0,0 +1,177 @@
+				Message:            "Fetched revision: main/b8e362c206e3d0cbb7ed22ced771a0056455a2fb",
+			},
+		},
+		Artifact: &sourcev1.Artifact{
+			Path:           "gitrepository/flux-system/test-tf-controller/b8e362c206e3d0cbb7ed22ced771a0056455a2fb.tar.gz",
+			URL:            server.URL() + "/file.tar.gz",
+			Revision:       "main/b8e362c206e3d0cbb7ed22ced771a0056455a2fb",
+			Digest:         "sha256:80ddfd18eb96f7d31cadc1a8a5171c6e2d95df3f6c23b0ed9cd8dddf6dba1406",
+			LastUpdateTime: metav1.Time{Time: updatedTime},
+		},
+	}
+
+	It("should be updated successfully.")
+	g.Expect(k8sClient.Status().Update(ctx, &testRepo)).Should(Succeed())
+
+	By("checking that the status and its URL gets reconciled.")
+	gitRepoKey := types.NamespacedName{Namespace: "flux-system", Name: sourceName}
+	createdRepo := &sourcev1.GitRepository{}
+	g.Expect(k8sClient.Get(ctx, gitRepoKey, createdRepo)).Should(Succeed())
+
+	Given("a Terraform resource with auto approve, without the list of outputs, attached to the given GitRepository resource")
+	By("creating a new TF resource and attaching to the repo via `sourceRef`.")
+	By("not specifying the outputs list of .spec.writeOutputsToSecret.")
+	helloWorldTF := infrav1.Terraform{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      terraformName,
+			Namespace: "flux-system",
+		},
+		Spec: infrav1.TerraformSpec{
+			PlanOnly: true,
+			Path:     "./terraform-hello-world-example",
+			SourceRef: infrav1.CrossNamespaceSourceReference{
+				Kind:      "GitRepository",
+				Name:      sourceName,
+				Namespace: "flux-system",
+			},
+			Interval:          metav1.Duration{Duration: time.Second * 10},
+			StoreReadablePlan: "human",
+			WriteOutputsToSecret: &infrav1.WriteOutputsToSecretSpec{
+				Name: "tf-output-" + terraformName,
+				// NOTE comment out only. Please not remove this line: Outputs: []string{},
+			},
+		},
+	}
+	It("should be created and attached successfully.")
+	g.Expect(k8sClient.Create(ctx, &helloWorldTF)).Should(Succeed())
+	defer func() { g.Expect(k8sClient.Delete(ctx, &helloWorldTF)).Should(Succeed()) }()
+
+	By("checking that the TF resource existed inside the cluster.")
+	helloWorldTFKey := types.NamespacedName{Namespace: "flux-system", Name: terraformName}
+	createdHelloWorldTF := infrav1.Terraform{}
+	// We'll need to retry getting this newly created Terraform, Given that creation may not immediately happen.
+	g.Eventually(func() bool {
+		err := k8sClient.Get(ctx, helloWorldTFKey, &createdHelloWorldTF)
+		if err != nil {
+			return false
+		}
+		return true
+	}, timeout, interval).Should(BeTrue())
+
+	It("should be reconciled and produce the correct output secret.")
+	By("checking that the named output secret contains all outputs.")
+	outputKey := types.NamespacedName{Namespace: "flux-system", Name: "tfplan-default-" + terraformName}
+	planOutput := corev1.ConfigMap{}
+	g.Eventually(func() (int, error) {
+		err := k8sClient.Get(ctx, outputKey, &planOutput)
+		if err != nil {
+			return -1, err
```

Comment:
You already checked that it exists, above; if there's an error now, that would represent an unexpected outcome, i.e,. a failure. I think it's better to use `g.Expect(k8sClient.Get(...)).To(Succeed())` here.

You can just remove this `Eventually` call, and change the assertion below. What you want to check is "the object exists, and as soon as it exists, it has this content". The above call to Eventually checks it exists, and the below can directly assert the content of `planObject` -- there's no further things to wait for that will change the result.

Reply:
> You already checked that it exists, above;

Where? This is the first place where we get the `ConfigMap`. The `Eventually` is there to make sure it retries a few times and wait until the terraform controller creates the `ConfigMap`, it is created only after the plan output is available form the `terraform plan`.

This part can be moved out of the `Eventually` function:
```
return len(planOutput.Data), nil
```
But I don't think the `Eventually` can be removed (asked @chanwit, he may knows more if we need it or not).

Reply:
You are right, I got lost in the Eventuallys. You can discard the Eventually that follows and just assert the content of `planOutput`. (Unless I have misunderstood, and the plan object is created and populated later, by the controller).